core/src/cluster_info.rs

```diff
@@ -2089,56 +2080,42 @@ impl ClusterInfo {
                 retain_staked(resp, stakes);
             }
         }
-        let pull_responses: Vec<_> = pull_responses
-            .into_iter()
-            .zip(addrs.into_iter())
-            .filter(|(response, _)| !response.is_empty())
-            .collect();
-
-        if pull_responses.is_empty() {
-            return packets;
-        }
-
-        let stats: Vec<_> = pull_responses
+        let (responses, scores): (Vec<_>, Vec<_>) = addrs
             .iter()
-            .enumerate()
-            .map(|(i, (responses, _from_addr))| {
-                responses
-                    .iter()
-                    .enumerate()
-                    .map(|(j, response)| {
-                        let score = if stakes.contains_key(&response.pubkey()) {
-                            2
-                        } else {
-                            1
-                        };
-                        let score = match response.data {
-                            CrdsData::ContactInfo(_) => 2 * score,
-                            _ => score,
-                        };
-                        ResponseScore {
-                            to: i,
-                            responses_index: j,
-                            score,
-                        }
-                    })
-                    .collect::<Vec<ResponseScore>>()
+            .zip(pull_responses)
+            .flat_map(|(addr, responses)| repeat(addr).zip(responses))
+            .map(|(addr, response)| {
+                let age = now.saturating_sub(response.wallclock());
+                let score = DEFAULT_EPOCH_DURATION_MS
+                    .saturating_sub(age)
+                    .div(CRDS_GOSSIP_PULL_CRDS_TIMEOUT_MS)
```

Comment:
Double checking, are most Crds values expected to be refreshed within `CRDS_GOSSIP_PULL_CRDS_TIMEOUT_MS`, timeout?

Reply:
Also, any concerns that somebody could finegle their wallclock to be further ahead and have a higher score?

Reply:
> Also, any concerns that somebody could finegle their wallclock to be further ahead and have a higher score?

Good point. If the value's wallclock exceeds caller's wallclock, it will be filtered out:
https://github.com/solana-labs/solana/blob/2ae57c172/core/src/crds_gossip_pull.rs#L529

and, the `.div(CRDS_GOSSIP_PULL_CRDS_TIMEOUT_MS)` in this change will mask out small differences in wallclocks. So I think it should be fine. Nodes are just better off to use the right wallclock.

> are all Crds values expected to be refreshed within CRDS_GOSSIP_PULL_CRDS_TIMEOUT_MS timeout, or are there some that are refreshed less often

Not sure what you mean by refresh. Only, contact-info and node-instance are refreshed periodically:
https://github.com/solana-labs/solana/blob/2ae57c172/core/src/cluster_info.rs#L546-L549

Reply:
Oh I was mostly concerned if some older values from higher staked nodes would now be outprioritized by later messages by lower staked nodes. But I think you're right, the message from a staked validator would have to half an epoch old to be outprioritized  by a newer response from an unstaked node

Reply:
There is also a score bump for values from staked nodes.
Potentially we can increase that if needed:
https://github.com/solana-labs/solana/blob/e8b35a4f7/core/src/cluster_info.rs#L2102-L2106